src/output/gtk-shell.cpp

```diff
@@ -2,77 +2,205 @@
-static void handle_gtk_surface_destroy(wl_resource *resource) { }
 
+/**
+ * Modal dialogs may be handled differently than non-modal dialogs.
+ * It is a hint that this should be attached to the parent surface.
+ * In gnome this does not affect input-focus.
+ * This function sets the modal hint.
+ */
+static void handle_gtk_surface_set_modal(wl_client *client, wl_resource *resource)
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
+        view->store_data(std::unique_ptr<wf::custom_data_t>(), "gtk-shell-modal");
+}
+
+
+/**
+ * Modal dialogs may be handled differently than non-modal dialogs.
+ * It is a hint that this should be attached to the parent surface.
+ * In gnome this does not affect input-focus.
+ * This function removes the modal hint.
+ */
+static void handle_gtk_surface_unset_modal(wl_client *client, wl_resource *resource)
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
+        view->erase_data("gtk-shell-modal");
+}
+
+
+/**
+ * The surface requests focus, for example single instance applications like
+ * gnome-control-center, gnome-clocks, dconf-editor are single instance and if
+ * they are already running and launched again, this will request that they get focused.
+ * This function is superseeded by handle_gtk_surface_request_focus a newer equivalelent
+ * used by gtk-applications now. This function is for compatibility reasons.
+ */
+static void handle_gtk_surface_present(wl_client *client, wl_resource *resource, uint32_t time)
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
+    {
+        view_self_request_focus_signal data;
+        data.view = view;
+        view->get_output()->emit_signal("view-self-request-focus", &data);
+    }
+}
+
+
+/**
+ * The surface requests focus, for example single instance applications like
+ * gnome-control-center, gnome-clocks, dconf-editor are single instance and if
+ * they are already running and launched again, this will request that they get focused.
+ */
+static void handle_gtk_surface_request_focus(struct wl_client *client, struct wl_resource *resource, const char* startup_id)
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
+    {
+        view_self_request_focus_signal data;
+        data.view = view;
+        view->get_output()->emit_signal("view-self-request-focus", &data);
+    }
+}
+
+
+/**
+ * Destroys the gtk_surface object.
+ * TODO: May this be required for some kind of leak?
```

Comment:
```suggestion
```

There shouldn't be any leaks, because we are not allocating any objects when the surface is created aside from the wayland protocol object, and this is its destructor.

Reply:
Oh, I didn't realize that this is the destructor.